Sources/GRPCHTTP2Core/GRPCStreamStateMachine.swift

```diff
@@ -0,0 +1,687 @@
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is already open: shouldn't be sending metadata.")
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is closed: can't send metadata.")
+      }
+    }
+
+    mutating func send(message: [UInt8], endStream: Bool) throws {
+      // Client sends message.
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client not yet open.")
+      case .clientOpenServerIdle(var state):
+        state.framer.append(message)
+        if endStream {
+          self.state = .clientClosedServerIdle(.init(previousState: state))
+        } else {
+          self.state = .clientOpenServerIdle(state)
+        }
+      case .clientOpenServerOpen(var state):
+        state.framer.append(message)
+        if endStream {
+          self.state = .clientClosedServerOpen(.init(previousState: state))
+        } else {
+          self.state = .clientOpenServerOpen(state)
+        }
+      case .clientOpenServerClosed:
+        // The server has closed, so it makes no sense to send the rest of the request.
+        // Do nothing.
+        ()
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is closed, cannot send a message.")
+      }
+    }
+    
+    mutating func send(status: String, trailingMetadata: Metadata) throws {
+      throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client cannot send status and trailer.")
+    }
+    
+    /// Returns the client's next request to the server.
+    /// - Returns: The request to be made to the server.
+    mutating func nextOutboundMessage() throws -> ByteBuffer? {
+      switch self.state {
+      case .clientIdleServerIdle:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Client is not open yet.")
+      case .clientOpenServerIdle(var clientOpenState):
+        let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+        self.state = .clientOpenServerIdle(clientOpenState)
+        return request
+      case .clientOpenServerOpen(var clientOpenState):
+        let request = try clientOpenState.framer.next(compressor: clientOpenState.compressor)
+        self.state = .clientOpenServerOpen(clientOpenState)
+        return request
+      case .clientOpenServerClosed:
+        // Nothing to do: no point in sending request if server is closed.
+        return nil
+      case .clientClosedServerIdle, .clientClosedServerOpen, .clientClosedServerClosed:
+        throw assertionFailureAndCreateRPCErrorOnFailedPrecondition("Can't send request if client is closed.")
+      }
+    }
+    
+    mutating func receive(metadata: Metadata, endStream: Bool) throws -> OnMetadataReceived {
```

Comment:
Ah, I was thinking this conversion would happen in the handler. I kind of went back and forth on this though, wasn't sure if it fell more under the responsibilities of the handler or the state machine. I decided to assume the conversion would happen in the handler and thus the state machine would already get all the "correct" GRPC-specific types instead of types coming from the H2 layer.

Reply:
Ah right. I think it needs to do the conversion because it holds information required to do so (enabled compression algorithms being one example). It's already taking care of framing/deframing which is part of the conversion.